pkg/tools/closectx/context.go

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2021 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package closectx provides helper method to create Close/Unregister context in case of Request/Register failure
+package closectx
+
+import (
+	"context"
+	"time"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/clock"
+	"github.com/networkservicemesh/sdk/pkg/tools/extend"
+)
+
+const closeTimeout = 15 * time.Second
+
+// New creates a new context for Close/Unregister in case of Request/Register failure
+func New(ctx context.Context) (context.Context, context.CancelFunc) {
+	clockTime := clock.FromContext(ctx)
+
+	var timeout time.Duration
+	if deadline, ok := ctx.Deadline(); ok {
+		timeout = clockTime.Until(deadline)
+	}
+	if timeout < closeTimeout {
```

Comment:
This context is going to expire, if anything, *faster* the the initial context from which its taken... its not going to be appropriate for 'Have to call Close at some later point' ... the right answer here is closer to:

```go
func FromContext(ctx context.Context) func() context.Context {
	ctxFunc := func() context.Context { return context.Background() }
	if deadline, ok := ctx.Deadline(); ok {
		duration := time.Until(deadline)
		ctxFunc = func() context.Context {
			ctx, _ := context.WithTimeout(context.Background(), duration)
			return ctx
		}
	}
	return ctxFunc
}
```

Reply:
We increase `timeout` up to `closeTimeout` if it is less than that. So it does mean that new context is going to expire in the same time or slower than the initial context.
So using here pure `ctx.Deadline()` just leads us to the same old issue that should be fixed by `closectx`:
1. Request context expires (or is very close to be expired).
2. Error happens.
3. Chain element calls Close on the expired (almost expired) context.

Reply:
Not if you handle it like this:

```go
func (c *client) Request(ctx context.Context,request *networkservice.Request,opts...grpc.CallOptions) {
    ctxFunc := ctxfunc.FromContext(ctx)
    conn, err := next.Client(ctx).Request(ctx,request,opts...)
    ...
    err = doSomething(...)
    if err != nil {
        _ = c.Close(extend.WithValues(ctxFunc(),ctx),conn)
       return nil, err
    }
}
```

Reply:
You capture the time at the beginning of the Request, and utilize it at the end of the Request for cleanup... all time spent calling down chain and doing something is recovered.

Reply:
Looks great, done this in `postpone.Context()`.